lib/vector-config/src/component/marker.rs

```diff
@@ -0,0 +1,33 @@
+/// An enrichment table component.
+pub struct EnrichmentTableComponent;
+
+/// A provider component.
+pub struct ProviderComponent;
+
+/// A sink component.
+pub struct SinkComponent;
+
+/// A source component.
+pub struct SourceComponent;
+
+// A transform component.
+pub struct TransformComponent;
+
+// Marker trait representing a component.
+pub trait ComponentMarker: sealed::Sealed {}
+
+impl ComponentMarker for EnrichmentTableComponent {}
+impl ComponentMarker for ProviderComponent {}
+impl ComponentMarker for SinkComponent {}
+impl ComponentMarker for SourceComponent {}
+impl ComponentMarker for TransformComponent {}
+
+mod sealed {
+    pub trait Sealed {}
+
+    impl Sealed for super::EnrichmentTableComponent {}
+    impl Sealed for super::ProviderComponent {}
+    impl Sealed for super::SinkComponent {}
+    impl Sealed for super::SourceComponent {}
+    impl Sealed for super::TransformComponent {}
```

Comment:
The "sealed trait" pattern ensures that downstream crates can't implement `ComponentMarker` for themselves. This will eventually be more important when we start deriving implementations of traits such that you can't manually add a new variant to the `Sources` enum, etc, without having using the right macro which derives the correct component type-specific traits for you.